src/handshake.ts

```diff
@@ -1,89 +1,125 @@
     } else {
-      const receivedMessageBuffer = (await this.connection.readLP()).slice();
-      const plaintext = await this.xx.recvMessage(ns, decodeMessageBuffer(receivedMessageBuffer));
-    }
+      const receivedMessageBuffer = decodeMessageBuffer((await this.connection.readLP()).slice());
 
-    return ns;
+      const plaintext = await this.xx.recvMessage(this.session, receivedMessageBuffer);
+      logger("Stage 0 - Responder received proposed message and remote static public key.");
+    }
   }
 
   // stage 1
-  async exchange(isInitiator: boolean, session: NoiseSession): Promise<void> {
-    if (isInitiator) {
-      const receivedMessageBuffer = (await this.connection.readLP()).slice();
-      const plaintext = await this.xx.recvMessage(session, decodeMessageBuffer(receivedMessageBuffer));
+  async exchange(): Promise<void> {
+    if (this.isInitiator) {
+      const receivedMessageBuffer = decodeMessageBuffer((await this.connection.readLP()).slice());
+      const plaintext = await this.xx.recvMessage(this.session, receivedMessageBuffer);
+      logger('Stage 1 - Initiator received the message.');
     } else {
       // create payload as responder
-      const signedPayload = signPayload(this.staticKeys.privateKey, getHandshakePayload(this.staticKeys.publicKey));
-      const handshakePayload = await createHandshakePayload(this.remotePublicKey, signedPayload);
+      const signedPayload = signPayload(this.libp2pPrivateKey, getHandshakePayload(this.staticKeys.publicKey));
+      const handshakePayload = await createHandshakePayload(
+        this.libp2pPublicKey,
+        this.libp2pPrivateKey,
+        signedPayload,
+      );
 
-      const message = Buffer.concat([Buffer.alloc(0), handshakePayload]);
-      const messageBuffer = await this.xx.sendMessage(session, message);
+      const messageBuffer = await this.xx.sendMessage(this.session, handshakePayload);
       this.connection.writeLP(encodeMessageBuffer(messageBuffer));
+      logger('Stage 1 - Responder sent the message.')
     }
   }
 
   // stage 2
-  async finish(isInitiator: boolean, session: NoiseSession): Promise<void> {
-    if (isInitiator) {
-      const messageBuffer = await this.xx.sendMessage(session, Buffer.alloc(0));
+  async finish(): Promise<void> {
+    if (this.isInitiator) {
+      const messageBuffer = await this.xx.sendMessage(this.session, Buffer.alloc(0));
       this.connection.writeLP(encodeMessageBuffer(messageBuffer));
+      logger('Stage 2 - Initiator sent message.');
     } else {
       const receivedMessageBuffer = (await this.connection.readLP()).slice();
-      const plaintext = await this.xx.recvMessage(session, decodeMessageBuffer(receivedMessageBuffer));
+      const plaintext = await this.xx.recvMessage(this.session, decodeMessageBuffer(receivedMessageBuffer));
+      logger('Stage 2 - Responder received the message, finished handshake.')
+    }
+  }
+
+  encrypt(plaintext: bytes, session: NoiseSession): bytes {
```

Comment:
I'm pretty sure you won't get buffer from stream, it will most likely be Uint8Array so it might be good idea to adjust types and wrap it here using Buffer.from

Reply:
you mean to accept anything and explicitly convert?

Reply:
not anything, plaintext won't be bytes type, it will be Uint8Array